README.md
Leg Tracker
===========
for ROS Indigo

Usage
-------------------
- clone leg_tracker repo into your indigo catkin source directory
- install SciPy: 
    - $ sudo apt-get install python-scipy
- install pykalman (http://pykalman.github.io/#installation or https://pypi.python.org/pypi/pykalman): 
    - $ sudo pip install pykalman
- install munkres for Python:
    - $ sudo pip install munkres
- $ cd [your catkin_workspace]
- $ catkin_make
- $ roslaunch leg_tracker joint_leg_tracker.launch 


Demos
-------------------
- try $ roslaunch leg_tracker demo_stationary_simple_environment.launch to see a demo of it working on a stationary robot.
- there are a bunch more demo launch files in the /launch directory.
- video: [youtu.be/F2Qx3O5AyBs](youtu.be/F2Qx3O5AyBs), which also includes a brief explaination of the method. 


Subscribed topics
-------------------
- /scan
    - you can change the subscribed scan topic by changing the parameter "scan_topic" in the launch file.
- /tf
    - to get the transforms. You should specify the fixed frame by the parameter "fixed_frame" in the launch file.


Published topics
-------------------
- /people_tracked 
    - the positions and velocities of the tracked people as a PersonArray message (see: leg_tracker/msg/PersonArray.msg for the message specification).
- /visualization_marker
    - markers for the tracked people, other tracked objects and the centroids of the detected clusters.


Important parameters
-------------------
- scan_topic
- scan_frequency
    - right now, it's only been tuned for 7.5, 10 or 15Hz laser scanners. If you're using something faster than this, I would recommend downsampling to one of these frequencies.
- fixed_frame
- publish_people_frame
    - frame_id for /people_tracked messages. Default: fixed_frame.

Other parameters
-------------------
- max_detected_clusters
    - If the program is running slow because there are a lot of objects to be tracked, you can set this so it only tracks the "max_detected_clusters" closest objects. Default: -1 (i.e., unlimited).
- forest_file
    - where to find the configuration file for the OpenCV random forest which has been trained to determine the probability of a scan cluster being a human leg based on its shape.
- detection_threshold
    - Confidence threshold in leg shape to publish detect clusters. We found be results publishing and tracking all clusters, regardless of confidence in leg shape. Default: -1 (i.e., no threshold).
- cluster_dist_euclid
    - Euclidian distance for clustering of laser scan points. Any two laser scan points which are closer than this distance will be clustered together. If you change this, you might want to re-train the leg detector. Default: 0.13 (m).
- min_points_per_cluster
    - Minimum scan points for a cluster to be valid. If a cluster has less than this, then we assume it's noise and delete it. Default: 3.
- max_detect_distance
    - Maximum distance from laser scanner to consider detected clusters. Farther is better but decreases run speed. Default: 10.0 (m). 
- max_leg_pairing_dist
    - Maximum separation of legs. Legs further apart than this distance will not be matched. Default: 0.8 (m).
- confidence_threshold_to_maintain_track
    - How high the confidence needs to be of the current track's leg shape to initiate and maintain a person track. Higher values will yield lower false positive people tracks and lower values higher false negatives. Will likely need re-tuning if you retrain the leg detector. Default: 0.1. 
- publish_occluded
    - If people who weren't seen in the previous scan should be published to the /people_tracked topic. Default: True.
- use_scan_header_stamp_for_tfs
    - If you should use the scan headers to do tfs or just use the latest transform available. If you're trying to do repeatable tests on recorded bags than it might be useful to use the scan headers. Default: False (i.e., use latest tf available).
- display_detected_people
    - Previously used only for debugging.  Default: False.
- dist_travelled_together_to_initiate_leg_pair
    - How far a candidate leg pair must travel to initiate a person track. Default: 0.5 (m).
- in_free_space_threshold
    - A threshold to determine whether a track should be considered in free space or occupied space. Default: 0.06. 
- confidence_percentile
    - Confidence percentile for matching of clusters to tracks. Larger confidence percentile generates larger gates for the tracks to match to new clusters. You should be able to view the size of the gates in Rviz as circular visualization markers. Default: 0.9.
- max_std
    - Maximum standard deviation of covariance of a track before it is deleted: Default: 0.9.


**Compatible laser scanners**

- right now it's trained for laser scanners of 0.33 degree resolution. Different resolutions should still be runnable but would probably require retraining (see below) to get optimal performance.


Retraining the leg detector
-------------------
- gather positive examples by setting the scanner up in a populated area where you specify a bounding box (or alternatively, an arc), within which only peoples' legs will appear.
- record scan data in a rosbag.
- run extract_positive_training_clusters.launch on the recorded data to have it automatically annotate where the legs are in the scans. It will put a marker on any clusters that fall within the specified bounding box. You can check if the right clusters have been annotated by playing back the generated rosbag with the "--pause" option and stepping through it and visually checking in Rviz where the markers lie in the scan.
- gather negative examples by moving the scanner on a moble platform around empty rooms.
- run train_leg_detector and include the positive and negative examples as command line arguments (see the train_leg_detector.launch for an example). 
- a .yaml file will be output with the name of the "save_file" parameter in the train_leg_detector launch file.
- set the launch file for the joint_leg_tracker to load the .yaml file you produced.
- when running the joint_leg_tracker with your new detector, you will likely need to retune the "confidence_threshold_to_maintain_track" parameter so you get the desired tradeoff between false positive/false negative person tracks.


Under the hood 
-------------------
**Files for running**

- src/laser_processor.cpp - scan processing stuff. E.g., splitting scan into clusters based on euclidian distance.
- src/cluster_features.cpp - calculates geometric features of scan clusters. Used by detect_leg_clusters.
- src/detect_leg_clusters.cpp - detects leg clusters in the scan and publishes them to /detected_leg_clusters.
- scripts/joint_leg_tracker.py - uses detect_leg_clusters and a global nearest neighbour tracking method to track people. They're published as markers to rviz and to a people_tracked topic.
- scripts/individual_leg_tracker.py - alternative to joint_leg_tracker. I found better performance with the joint_leg_tracker.


** Files for retraining the detector **

- src/extract_positive_training_clusters.cpp - extracts positive example clusters from scans in a rosbag. They're saved to another rosbag to be used for training.
- src/train_leg_detector.cpp - train the random forest classifier using the positive example clusters and negative scans examples.


Speedups
-------------------
- I found using this version of munkres, which is based on numpy, sped up the data association matching a little bit: https://github.com/ben-eysenbach/munkres


Acknowledgement
-------------------
We used some code from the leg_detector package: http://wiki.ros.org/leg_detector. 
It does fundamentally the same task, but we found this version was more capable of tracking people longer, not mismatching people and producing less false positives in cases where a grid occupancy map is not provided a priori. See the reference paper and the leg_tracker_benchmarks repo for more details.


TODO
-------------------
- port tracker to C++ to improve speed
- use BFL's kalman filter in place of pykalman so users aren't required to install the external dependancy
- use a C++ version of munkres in place of python's munkres so users aren't required to install the external dependancy
- integrate a priori occupancy grid maps or SLAM gmapping maps to speed up detection and reduce false positives


Reference
-------------------
A. Leigh, J. Pineau, N. Olmedo and H. Zhang, Person Tracking and Following with 2D Laser Scanners, International Conference on Robotics and Automation (ICRA), Seattle, Washington, USA, 2015.

Citations appreciated!


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package leg_tracker
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.1 (2015-01-25)
------------------
* initial release

src/detect_leg_clusters.cpp
/*********************************************************************
* Software License Agreement (BSD License)
* 
*  Copyright (c) 2008, Willow Garage, Inc.
*  All rights reserved.
* 
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions
*  are met:
* 
*   * Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*   * Redistributions in binary form must reproduce the above
*     copyright notice, this list of conditions and the following
*     disclaimer in the documentation and/or other materials provided
*     with the distribution.
*   * Neither the name of the Willow Garage nor the names of its
*     contributors may be used to endorse or promote products derived
*     from this software without specific prior written permission.
* 
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*********************************************************************/
// ROS 
#include <ros/ros.h>

#include <tf/transform_listener.h>

#include <visualization_msgs/Marker.h>
#include <sensor_msgs/LaserScan.h>

// OpenCV
#include <opencv/cxcore.h>
#include <opencv/cv.h>
#include <opencv/ml.h>

// Local headers
#include <leg_tracker/laser_processor.h>
#include <leg_tracker/cluster_features.h>

// Custom messages
#include <leg_tracker/Leg.h>
#include <leg_tracker/LegArray.h>



/**
* @brief Detects clusters in laser scan with leg-like shapes
*/
class DetectLegClusters
{
public:
  /**
  * @brief Constructor
  */
  DetectLegClusters():
  scan_num_(0),
  num_prev_markers_published_(0)
  {  
    // Get ROS parameters  
    std::string forest_file;
    std::string scan_topic;
    if (!nh_.getParam("forest_file", forest_file))
      ROS_ERROR("ERROR! Could not get random forest filename");
    nh_.param("scan_topic", scan_topic, std::string("scan"));
    nh_.param("fixed_frame", fixed_frame_, std::string("odom"));
    nh_.param("detection_threshold", detection_threshold_, -1.0);
    nh_.param("cluster_dist_euclid", cluster_dist_euclid_, 0.13);
    nh_.param("min_points_per_cluster", min_points_per_cluster_, 3);                
    nh_.param("max_detect_distance", max_detect_distance_, 10.0);   
    nh_.param("marker_display_lifetime", marker_display_lifetime_, 0.2);   
    nh_.param("use_scan_header_stamp_for_tfs", use_scan_header_stamp_for_tfs_, false);
    nh_.param("max_detected_clusters", max_detected_clusters_, -1);

    // Print back
    ROS_INFO("forest_file: %s", forest_file.c_str());
    ROS_INFO("scan_topic: %s", scan_topic.c_str());
    ROS_INFO("fixed_frame: %s", fixed_frame_.c_str());
    ROS_INFO("detection_threshold: %.2f", detection_threshold_);
    ROS_INFO("cluster_dist_euclid: %.2f", cluster_dist_euclid_);
    ROS_INFO("min_points_per_cluster: %d", min_points_per_cluster_);
    ROS_INFO("max_detect_distance: %.2f", max_detect_distance_);    
    ROS_INFO("marker_display_lifetime: %.2f", marker_display_lifetime_);
    ROS_INFO("use_scan_header_stamp_for_tfs: %d", use_scan_header_stamp_for_tfs_);    
    ROS_INFO("max_detected_clusters: %d", max_detected_clusters_);    

    // Load random forst
    forest.load(forest_file.c_str());
    feat_count_ = forest.get_active_var_mask()->cols;

    latest_scan_header_stamp_with_tf_available_ = ros::Time::now();

    // ROS subscribers + publishers
    scan_sub_ =  nh_.subscribe(scan_topic, 10, &DetectLegClusters::laserCallback, this);
    markers_pub_ = nh_.advertise<visualization_msgs::Marker>("visualization_marker", 20);
    detected_leg_clusters_pub_ = nh_.advertise<leg_tracker::LegArray>("detected_leg_clusters",20);
  }

private:
  tf::TransformListener tfl_;

  CvRTrees forest;

  int feat_count_;

  ClusterFeatures cf_;

  int scan_num_;
  bool use_scan_header_stamp_for_tfs_;
  ros::Time latest_scan_header_stamp_with_tf_available_;

  ros::NodeHandle nh_;
  ros::Publisher markers_pub_;
  ros::Publisher detected_leg_clusters_pub_;
  ros::Subscriber scan_sub_;

  std::string fixed_frame_;
  
  double detection_threshold_;
  double cluster_dist_euclid_;
  int min_points_per_cluster_;  
  double max_detect_distance_;
  double marker_display_lifetime_;
  int max_detected_clusters_;

  int num_prev_markers_published_;


  /**
  * @brief Clusters the scan according to euclidian distance, 
  *        predicts the confidence that each cluster is a human leg and publishes the results
  * 
  * Called every time a laser scan is published.
  */
  void laserCallback(const sensor_msgs::LaserScan::ConstPtr& scan)
  {         
    laser_processor::ScanProcessor processor(*scan); 
    processor.splitConnected(cluster_dist_euclid_);        
    processor.removeLessThan(min_points_per_cluster_);    

    // OpenCV matrix needed to use the OpenCV random forest classifier
    CvMat* tmp_mat = cvCreateMat(1, feat_count_, CV_32FC1); 
    
    leg_tracker::LegArray detected_leg_clusters;
    detected_leg_clusters.header.frame_id = scan->header.frame_id;
    detected_leg_clusters.header.stamp = scan->header.stamp;

    // Find out the time that should be used for tfs
    bool transform_available;
    ros::Time tf_time;
    // Use time from scan header
    if (use_scan_header_stamp_for_tfs_)
    {
      tf_time = scan->header.stamp;

      try
      {
        tfl_.waitForTransform(fixed_frame_, scan->header.frame_id, tf_time, ros::Duration(1.0));
        transform_available = tfl_.canTransform(fixed_frame_, scan->header.frame_id, tf_time);
      }
      catch(tf::TransformException ex)
      {
        ROS_INFO("Detect_leg_clusters: No tf available");
        transform_available = false;
      }
    }
    else
    {
      // Otherwise just use the latest tf available
      tf_time = ros::Time(0);
      transform_available = tfl_.canTransform(fixed_frame_, scan->header.frame_id, tf_time);
    }
    
    // Store all processes legs in a set ordered according to their relative distance to the laser scanner
    std::set <leg_tracker::Leg, CompareLegs> leg_set;
    if (!transform_available)
    {
      ROS_INFO("Not publishing detected leg clusters because no tf was available");
    }
    else // transform_available
    {
      // Iterate through all clusters
      for (std::list<laser_processor::SampleSet*>::iterator cluster = processor.getClusters().begin();
       cluster != processor.getClusters().end();
       cluster++)
      {   
        // Get position of cluster in laser frame
        tf::Stamped<tf::Point> position((*cluster)->getPosition(), tf_time, scan->header.frame_id);
        float rel_dist = pow(position[0]*position[0] + position[1]*position[1], 1./2.);
        
        // Only consider clusters within max_distance. 
        if (rel_dist < max_detect_distance_)
        {
          // Classify cluster using random forest classifier
          std::vector<float> f = cf_.calcClusterFeatures(*cluster, *scan);
          for (int k = 0; k < feat_count_; k++)
            tmp_mat->data.fl[k] = (float)(f[k]);
          float probability_of_leg = forest.predict_prob(tmp_mat);

          // Consider only clusters that have a confidence greater than detection_threshold_                 
          if (probability_of_leg > detection_threshold_)
          { 
            // Transform cluster position to fixed frame
            // This should always be succesful because we've checked earlier if a tf was available
            bool transform_successful_2;
            try
            {
              tfl_.transformPoint(fixed_frame_, position, position);
              transform_successful_2 = true;
            }
            catch (tf::TransformException ex)
            {
              ROS_ERROR("%s",ex.what());
              transform_successful_2 = false;
            }

            if (transform_successful_2)
            {  
              // Add detected cluster to set of detected leg clusters, along with its relative position to the laser scanner
              leg_tracker::Leg new_leg;
              new_leg.position.x = position[0];
              new_leg.position.y = position[1];
              new_leg.confidence = probability_of_leg;
              leg_set.insert(new_leg);
            }
          }
        }
      }     
    }    
 

    // Publish detected legs to /detected_leg_clusters and to rviz
    // They are ordered from closest to the laser scanner to furthest  
    int clusters_published_counter = 0;
    int id_num = 1;      
    for (std::set<leg_tracker::Leg>::iterator it = leg_set.begin(); it != leg_set.end(); ++it)
    {
      // Publish to /detected_leg_clusters topic
      leg_tracker::Leg leg = *it;
      detected_leg_clusters.legs.push_back(leg);
      clusters_published_counter++;

      // Publish marker to rviz
      visualization_msgs::Marker m;
      m.header.stamp = scan->header.stamp;
      m.header.frame_id = fixed_frame_;
      m.ns = "LEGS";
      m.id = id_num++;
      m.type = m.SPHERE;
      m.pose.position.x = leg.position.x ;
      m.pose.position.y = leg.position.y;
      m.pose.position.z = 0.2;
      m.scale.x = 0.13;
      m.scale.y = 0.13;
      m.scale.z = 0.13;
      m.color.a = 1;
      m.color.r = 0;
      m.color.g = leg.confidence;
      m.color.b = leg.confidence;
      markers_pub_.publish(m);

      // Comparison using '==' and not '>=' is important, as it allows <max_detected_clusters_>=-1 
      // to publish infinite markers
      if (clusters_published_counter == max_detected_clusters_) 
        break;
    }

    // Clear remaining markers in Rviz
    for (int id_num_diff = num_prev_markers_published_-id_num; id_num_diff > 0; id_num_diff--)
    {
      visualization_msgs::Marker m;
      m.header.stamp = scan->header.stamp;
      m.header.frame_id = fixed_frame_;
      m.ns = "LEGS";
      m.id = id_num_diff + id_num;
      m.action = m.DELETE;
      markers_pub_.publish(m);
    }
    num_prev_markers_published_ = id_num; // For the next callback

    detected_leg_clusters_pub_.publish(detected_leg_clusters);
    cvReleaseMat(&tmp_mat);
  }


  /**
  * @brief Comparison class to order Legs according to their relative distance to the laser scanner
  */
  class CompareLegs
  {
  public:
      bool operator ()(const leg_tracker::Leg &a, const leg_tracker::Leg &b)
      {
          float rel_dist_a = pow(a.position.x*a.position.x + a.position.y*a.position.y, 1./2.);
          float rel_dist_b = pow(b.position.x*b.position.x + b.position.y*b.position.y, 1./2.);          
          return rel_dist_a < rel_dist_b;
      }
  };

};





int main(int argc, char **argv)
{
  ros::init(argc, argv, "detect_leg_clusters");
  DetectLegClusters dlc;
  ros::spin();
  return 0;
}

